### PR TITLE
Implementation of Event System

### PR DESCRIPTION
**Pull request by Joakim-Holmberg**
# Additions
In order to implement the event system the following things have been added:
- New Event class. The event contains a predicate and a list of actions. The ruleset now contains a set of events. Whenever a move is made we loop through all events and evaluate their predicate over the transition that was just made. If the predicate holds, the event's list of actions are performed on the board.
- A method that constructs a promotion Event for pawns. This Event will replace a pawn with a queen if it is on the last rank.
- A method that constructs an "explosion" Event, used in the implementation of atomic chess.
- An implementation of atomic chess in the GameFactory. This is just for fun, check out this link for more info: https://en.wikipedia.org/wiki/Atomic_chess
- PositionHasRank predicate with tests, used to determine if a position has a certain rank (needed for the promotion event).
- ActionSetPiece that sets the piece at the internal position to the internal pieceIdentifier.
# Changes
Some changes that have been made in this pull request:
- All predicates that evaluate positions now use a new enum called RelativeTo. It is used to determine whether relative positions should be calculated relative to the moveFrom or moveTo variable of the given BoardTransition when the predicate is evaluated.

### CODE REVIEW

File: ChessVariantsLogic/Rules/Event.cs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    /// <summary>
+    /// Evaluates if the internal predicate holds for the given BoardTransition.
+    /// </summary>
+    /// <returns>True if the predicate holds, otherwise false.</returns>
+    public bool ShouldRun(BoardTransition lastTransition)
+    {
+        return _predicate.Evaluate(lastTransition);
+    }
```
Comment (sjohan99) [inline ChessVariantsLogic/Rules/Event.cs:31]: Parameters missing from documentation

File: ChessVariantsLogic/Rules/Event.cs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+        // Here we utilize the Move class to avoid code repetition.
+        // This is almost a bit of a hack though, so it might have to be changed in the future.
+
+        string move = _relativeTo == RelativeTo.FROM ? from + from : to + to;
+
+        Move moveEvent = new Move(_actions, move, lastMove.PieceClassifier);
+
+        return moveEvent.Perform(moveWorker);
+    }
```
Comment (sjohan99) [inline ChessVariantsLogic/Rules/Event.cs:53]: Return documentation is wrong. Also no whitespace between docs and code.

File: ChessVariantsLogic/Rules/Event.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+    /// <summary>
+    /// Runs the event on the given <paramref name="moveWorker"/>. This does not take into account whether the event actually should be run.
+    /// </summary>
+    /// <param name="moveWorker">The MoveWorker we want to run the event on.</param>
+    /// <returns>True if the predicate holds, otherwise false.</returns>
+
+    public GameEvent Run(MoveWorker moveWorker)
+    {
+        Move lastMove = moveWorker.getLastMove();
```
Comment (sjohan99) [inline ChessVariantsLogic/Rules/Event.cs:41]: Assigns nullable type to non-nullable type

File: ChessVariantsLogic/Rules/Event.cs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+        var (from, to) = MoveWorker.ParseMove(lastMove.FromTo);
```
Comment (sjohan99) [inline ChessVariantsLogic/Rules/Event.cs:43]: Unhandled null possibility

**Review by sjohan99 — CHANGES_REQUESTED** [review]
Good work! I pointed out a few very small issues, other than that it looks good! :D

**Comment by Joakim-Holmberg** [comment]
Yes, all these issues have now been resolved.

File: ChessVariantsLogic/Rules/Event.cs (hunk 1 above)
Comment (Joakim-Holmberg) [reply ChessVariantsLogic/Rules/Event.cs]: Fixed now.

File: ChessVariantsLogic/Rules/Event.cs (hunk 2 above)
Comment (Joakim-Holmberg) [reply ChessVariantsLogic/Rules/Event.cs]: Fixed now.

File: ChessVariantsLogic/Rules/Event.cs (hunk 3 above)
Comment (Joakim-Holmberg) [reply ChessVariantsLogic/Rules/Event.cs]: Fixed now.

**Review by sjohan99 — APPROVED** [review]
Looks awful...LY GOOD

**Pull request merged**